### PR TITLE
#2443: VerifyMojo

### LINKED ISSUE #2443: VerifyMojo.java:35-40: Implement VerifyMojo. VerifyMojo...
The puzzle `1708-536ee550` from #1708 has to be resolved: 

https://github.com/objectionary/eo/blob/16e21ff21bd5ee1560ae385b650734f40e868fce/eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java#L35-L40

The puzzle was created by rultor on 25-Aug-23. 

Estimate: 30 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by levBagryansky**
Closes #2443

---

### Detailed summary
- Added a new `VERIFIED` attribute to the `ForeignTojos` class to store the path to the verified xmir file.
- Updated the `verify.groovy` script to use the new directory structure.
- Added a new `verify` goal to the `pom.xml` file.
- Updated the `TranspileMojo` class to use the `verified` method instead of `shaken`.
- Added a new `VerifyMojo` class to perform verification of XMIR programs.
- Added tests for the `VerifyMojo` class.

> The following files were skipped due to too many changes: `eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java`, `eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by maxonfjvipon** [comment]
@levBagryansky it was suggested [here](https://github.com/objectionary/eo/issues/1708) to add `verify` step after `assemble` step, not during

**Comment by levBagryansky** [comment]
@maxonfjvipon could you pleas check it again

File: eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java [hunk 1]
```diff
@@ -115,4 +119,77 @@ void detectsCriticalError(@TempDir final Path temp) {
...
+                .execute(new FakeMaven.Verify())
+        );
+    }
+
+    /**
+     * Apply XSL transformation.
+     * @param xsl Path to XSL within classpath
+     * @param xml Path to XML to be tranformed
+     */
+    private void applyXsl(final String xsl, final Path xml) throws Exception {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java:184]: @levBagryansky can we make this method static?

File: eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java [hunk 2]
```diff
@@ -115,4 +119,77 @@ void detectsCriticalError(@TempDir final Path temp) {
             "Program with duplicate names should have failed on critical error, but it didn't"
         );
     }
+
+    @Test
+    void failsOptimization(@TempDir final Path temp) {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java:124]: @levBagryansky let's add assertion messages here and for all the tests below

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java [hunk 3]
```diff
@@ -23,28 +23,45 @@
...
+     * Whether we should fail on error.
+     *
+     * @checkstyle MemberNameCheck (7 lines)
+     * @since 0.34.0
+     */
+    @SuppressWarnings("PMD.ImmutableField")
+    @Parameter(
+        property = "eo.failOnError",
+        defaultValue = "true")
+    private boolean failOnError = true;
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java:63]: @levBagryansky the main idea of the VerifyMojo was to get rid of `failOnError` flag. More details [here](https://github.com/objectionary/eo/issues/1708). Long story short - if there's at least one ERROR - throw an exception. If there's at least on WARNING - throw an exception if `failOnWarning` is true

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java [hunk 4]
```diff
@@ -23,28 +23,45 @@
...
- *  mojo is implemented - need to remove "failOnError" flag from
- *  {@link OptimizeMojo} and put "verify" step right after "assemble" in all
- *  pom.xml files
  */
 @Mojo(
     name = "verify",
     defaultPhase = LifecyclePhase.PROCESS_SOURCES,
     threadSafe = true
 )
 public final class VerifyMojo extends SafeMojo {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java:51]: @levBagryansky I think you've lost `failOnWarning` flag in the class

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]
@levBagryansky good one, just several comments and suggestions from my side

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java (hunk 4 above)
Comment (levBagryansky) [reply eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java:51]: @maxonfjvipon it is already here

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java (hunk 3 above)
Comment (levBagryansky) [reply eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java]: @maxonfjvipon done

File: eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java (hunk 1 above)
Comment (levBagryansky) [reply eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java]: @maxonfjvipon done.

**Review by maxonfjvipon — APPROVED** [review]
@levBagryansky LGTM! Thanks.
@volodya-lombrozo WDYT?

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java [hunk 5]
```diff
@@ -60,11 +65,51 @@ public final class VerifyMojo extends SafeMojo {
...
+                ),
+                ForeignTojo::withVerified,
+                ForeignTojo::shaken
             )
+        ).count();
+        if (total > 0) {
+            Logger.info(
+                this,
+                "Verified %d out of %d XMIR program(s)", total,
+                tojos.size()
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java:92]: @levBagryansky You can get `int all = tojos.size()` once and in this case you won't need `tojos` variable.

File: eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java [hunk 6]
```diff
@@ -100,19 +105,58 @@ void doesNotDetectWarningWithoutCorrespondingFlag(@TempDir final Path temp) {
     }
 
     @Test
-    @Disabled
-    void detectsCriticalError(@TempDir final Path temp) {
+    void failsOptimizationOnError(@TempDir final Path temp) {
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java:108]: @levBagryansky Could you add human readable message for this assertion, please?
https://www.yegor256.com/2023/01/19/layout-of-tests.html#assertions

File: eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java [hunk 7]
```diff
@@ -100,19 +105,58 @@ void doesNotDetectWarningWithoutCorrespondingFlag(@TempDir final Path temp) {
...
+                "  [] > @",
+                "    hello > test"
+            );
+        VerifyMojoTest.applyXsl(
+            "org/eolang/maven/set-warning-severity.xsl",
+            maven.execute(ParseMojo.class)
+                .result()
+                .get("target/1-parse/foo/x/main.xmir")
+        );
+        Assertions.assertThrows(
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/test/java/org/eolang/maven/VerifyMojoTest.java:139]: @levBagryansky Could you add human readable message for this assertion, please?
https://www.yegor256.com/2023/01/19/layout-of-tests.html#assertions

File: eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java (hunk 5 above)
Comment (levBagryansky) [reply eo-maven-plugin/src/main/java/org/eolang/maven/VerifyMojo.java:92]: @volodya-lombrozo `tojos` is used previously too

**Review by volodya-lombrozo — APPROVED** [review]
@levBagryansky Looks good to me. Thanks.

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**